cmd/collect-profiles/main.go

```diff
@@ -0,0 +1,364 @@
+
+			// Attempt to delete all but the newest configMaps generated by this job
+			errs := []error{}
+			for _, cm := range expiredConfigMaps {
+				if err := cfg.Client.Delete(cmd.Context(), &cm); err != nil {
+					errs = append(errs, err)
+					continue
+				}
+				klog.Infof("Successfully deleted configMap %s/%s", cm.GetNamespace(), cm.GetName())
+			}
+
+			// If a delete call failed, abort to avoid creating new configMaps
+			if len(errs) != 0 {
+				return fmt.Errorf("error deleting expired pprof configMaps: %v", errs)
+			}
+
+			certPath := filepath.Join(certMountPath, corev1.TLSCertKey)
+			keyPath := filepath.Join(certMountPath, corev1.TLSPrivateKeyKey)
+
+			if err := verifyCertAndKeyExist(certPath, keyPath); err != nil {
+				klog.Infof("error verifying provided cert and key: %v", err)
+				klog.Info("generating a new cert and key")
+				return populateServingCert(cmd.Context(), cfg.Client)
+			}
+
+			httpClient, err := getHttpClient(certPath, keyPath)
+			if err != nil {
+				return err
+			}
+
+			// Track successfully created configMaps by generateName for each endpoint being scrapped.
+			createdCM := map[string]struct{}{}
+
+			for _, a := range validatedArguments {
+				b, err := requestURLBody(httpClient, a.url)
+				if err != nil {
+					klog.Infof("error retrieving pprof profile: %v", err)
+					continue
+				}
+
+				cm := &corev1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						GenerateName: a.generateName,
+						Namespace:    namespace,
+						Labels: map[string]string{
+							profileConfigMapLabelKey: "",
+						},
+					},
+					Immutable: getTruePointer(),
+					BinaryData: map[string][]byte{
+						"profile.pb.gz": b,
+					},
+				}
+
+				if err := cfg.Client.Create(cmd.Context(), cm); err != nil {
+					klog.Errorf("error created configMap %s/%s: %v", cm.GetNamespace(), cm.GetName(), err)
+					continue
+				}
+
+				klog.Infof("Successfully created configMap %s/%s", cm.GetNamespace(), cm.GetName())
+				createdCM[a.generateName] = struct{}{}
+			}
+
+			// Delete the configMaps which are no longer the newest
+			for _, cm := range newestConfigMaps {
+				// Don't delete ConfigMaps that were not replaced
+				// Also prevents deletes of configMaps with generateNames not included in command
+				if _, ok := createdCM[cm.GenerateName]; !ok {
+					continue
+				}
+				if err := cfg.Client.Delete(cmd.Context(), &cm); err != nil {
+					errs = append(errs, err)
+					continue
+				}
+				klog.Infof("Successfully deleted configMap %s/%s", cm.GetNamespace(), cm.GetName())
+			}
+
+			if len(errs) != 0 {
+				return fmt.Errorf("error deleting existing pprof configMaps: %v", errs)
+			}
+
+			// Update serving cert after a successful run
+			return populateServingCert(cmd.Context(), cfg.Client)
+		},
+	}
+}
+
+func verifyCertAndKeyExist(certPath, keyPath string) error {
+	fi, err := os.Stat(certPath)
+	if err != nil {
+		return err
+	}
+	if fi.Size() == 0 {
```

Comment:
Trying to wrap my head around when a file would exist but contain zero bytes. A `$ touch <file>` kind of operation?

Reply:
It's a bit of a chicken and an egg problem.

The service-ca operator cannot generate certs for use with client authentication. We elected to have the CronJob generate a self-signed cert and rotate the cert with each run. The tls secret containing the self-signed cert must be mounted to the pod, so it has to exist before the pod can populate it. You cannot omit the `tls.key` or `tls.crt` values from the data of a TLS Secret, so they had to be set to "". The initial run of the CronJob will have two empty files names `tls.key` and `tls.crt`.

The entire point of the above is to handle that initial run, update the secret, exit gracefully, OLM/Catlog Operator Deployments will then notice the updated secret and use that for the client, and allow subsequent CronJob runs to use that secret to store pprof data.